### PR TITLE
Shivamvohra/blobs

### PR DESCRIPTION
**Pull request by shivam-vohra**
functionality for blob storage methods. test functions are defined as well for integration testing.

### CODE REVIEW

File: backend/src/storage/blob.rs [hunk 1]
```diff
@@ -52,57 +130,304 @@ impl StorageProvider for BlobStorageProvider {
...
         Ok(Some(flag))
     }
 
-    async fn create_bool_flag(&self, _flag: BoolFlag) -> Result<(), StorageError> {
-        todo!()
+    async fn create_bool_flag(&self, flag: BoolFlag) -> Result<(), StorageError> {
+        if self.get_bool_flag(&flag.id).await?.is_some() {
+            return Err(StorageError::AlreadyExists);
+        }
+        println!("Checking if exists");
```
Comment (trstruth) [inline backend/src/storage/blob.rs]: either remove or comment all `println!`s, we'll replace these with logger calls in the future

File: backend/src/storage/blob.rs [hunk 2]
```diff
@@ -52,57 +130,304 @@ impl StorageProvider for BlobStorageProvider {
...
+        if self.name_exists(&flag.name).await? {
+            return Err(StorageError::AlreadyExists);
+        }
+        println!("Checked exists");
```
Comment (trstruth) [inline backend/src/storage/blob.rs]: same as above

File: backend/src/storage/blob.rs [hunk 3]
```diff
@@ -52,57 +130,304 @@ impl StorageProvider for BlobStorageProvider {
...
+
+        let res = blob_client
+            .put_block_blob(flag_str.clone())
+            .content_type("text/plain")
+            .await
+            .map_err(|e| {
+                StorageError::DatabaseError(format!("Failed to write flag data to blob: {}", e))
+            })?;
+
+        println!("1-put_block_blob {res:?}");
```
Comment (trstruth) [inline backend/src/storage/blob.rs]: same as above

File: backend/src/storage/blob.rs [hunk 4]
```diff
@@ -52,57 +130,304 @@ impl StorageProvider for BlobStorageProvider {
...
+        }
+
+        if self.name_exists(&flag.name).await? {
+            return Err(StorageError::AlreadyExists);
+        }
+
+        let blob_client = self.container_client.blob_client(flag.id.clone());
+
+        let flag_str = serde_json::to_string(&flag).map_err(|e| {
+            StorageError::SerializationError(format!("failed to parse json: {}", e))
```
Comment (trstruth) [inline backend/src/storage/blob.rs:332]: nit: can we make the message "failed to serialize json"

File: backend/src/storage/blob.rs [hunk 5]
```diff
@@ -52,57 +130,304 @@ impl StorageProvider for BlobStorageProvider {
...
+        while let Some(value) = stream.next().await {
+            let data = value
+                .map_err(|e| StorageError::DatabaseError(format!("failed to read stream: {}", e)))?
+                .data
+                .collect()
+                .await
+                .map_err(|e| {
+                    StorageError::DatabaseError(format!("failed to collect stream: {}", e))
+                })?;
+            println!("received {:?} bytes", data.len());
```
Comment (trstruth) [inline backend/src/storage/blob.rs]: same as above

File: backend/src/storage/blob.rs [hunk 6]
```diff
@@ -50,59 +155,301 @@ impl StorageProvider for BlobStorageProvider {
...
 
-    async fn create_bool_flag(&self, _flag: BoolFlag) -> Result<(), StorageError> {
-        todo!()
+    async fn create_bool_flag(&self, flag: BoolFlag) -> Result<(), StorageError> {
+        if self.get_bool_flag(&flag.id).await?.is_some() {
+            return Err(StorageError::AlreadyExists);
+        }
+        if self.name_exists(&flag.name).await? {
+            return Err(StorageError::AlreadyExists);
+        }
```
Comment (trstruth) [inline backend/src/storage/blob.rs:185]: kinda feels like this is a duplicate check, I think the metadata file should only be used to look up a flag by name

File: backend/src/storage/blob.rs [hunk 7]
```diff
@@ -50,59 +155,301 @@ impl StorageProvider for BlobStorageProvider {
...
+
+        let blob_client = self.container_client.blob_client(flag.id.clone());
+        
+        let flag_str = serde_json::to_string(&flag).map_err(|e| {
+            StorageError::SerializationError(format!("failed to serialize json: {}", e))
+        })?;
+
+        let _res = blob_client
```
Comment (trstruth) [inline backend/src/storage/blob.rs]: looks like we don't do anything with `_res`, can we remove the binding ?

File: backend/src/storage/blob.rs (hunk 6 above)
Comment (shivam-vohra) [reply backend/src/storage/blob.rs:185]: the first check makes sure the id isnt there, but then the second check makes sure the name isnt there. the get_bool_flag method doesnt call the metadata.

File: backend/src/storage/blob.rs (hunk 7 above)
Comment (shivam-vohra) [reply backend/src/storage/blob.rs]: yes will adjust

File: backend/src/storage/blob.rs (hunk 6 above)
Comment (trstruth) [reply backend/src/storage/blob.rs:185]: sync'd with shiv, leaving this in prevents flags with duplicate names being created

**Pull request merged**